api/client/mfa_test.go

```diff
@@ -0,0 +1,160 @@
+*/
+
+package client
+
+import (
+	"context"
+	"testing"
+
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/gravitational/trace"
+	"github.com/stretchr/testify/require"
+	"google.golang.org/protobuf/types/known/emptypb"
+
+	"github.com/gravitational/teleport/api/client/proto"
+	"github.com/gravitational/teleport/api/metadata"
+	"github.com/gravitational/teleport/api/types"
+)
+
+type mfaService struct {
+	*proto.UnimplementedAuthServiceServer
+}
+
+func (s *mfaService) Ping(ctx context.Context, req *proto.PingRequest) (*proto.PingResponse, error) {
+	return &proto.PingResponse{}, nil
+}
+
+func (s *mfaService) CreateAuthenticateChallenge(ctx context.Context, req *proto.CreateAuthenticateChallengeRequest) (*proto.MFAAuthenticateChallenge, error) {
+	return &proto.MFAAuthenticateChallenge{}, nil
+}
+
+const otpTestCode = "otp-test-code"
+
+func (s *mfaService) CreateUser(ctx context.Context, req *types.UserV2) (*emptypb.Empty, error) {
+	mfaResp, err := metadata.MFACredentialsFromContext(ctx)
+	if trace.IsBadParameter(err) {
+		return nil, trace.Wrap(ErrAdminActionMFARequired)
+	} else if err != nil {
+		return nil, trace.Wrap(ErrAdminActionMFARequired, "failed to retrieve MFA credentials from context with error: %v", err)
+	}
+
+	switch r := mfaResp.Response.(type) {
+	case *proto.MFAAuthenticateResponse_TOTP:
+		if r.TOTP.Code != otpTestCode {
+			return nil, trace.AccessDenied("failed MFA verification")
+		}
+	default:
+		return nil, trace.BadParameter("unexpected mfa response type %T", r)
+	}
+
+	return &empty.Empty{}, nil
+}
+
+func (s *mfaService) GetUsers(req *proto.GetUsersRequest, stream proto.AuthService_GetUsersServer) error {
+	mfaResp, err := metadata.MFACredentialsFromContext(stream.Context())
+	if trace.IsBadParameter(err) {
+		return trace.Wrap(ErrAdminActionMFARequired)
+	} else if err != nil {
+		return trace.Wrap(ErrAdminActionMFARequired, "failed to retrieve MFA credentials from context with error: %v", err)
+	}
+
+	switch r := mfaResp.Response.(type) {
+	case *proto.MFAAuthenticateResponse_TOTP:
+		if r.TOTP.Code != otpTestCode {
+			return trace.AccessDenied("failed MFA verification")
+		}
+	default:
+		return trace.BadParameter("unexpected mfa response type %T", r)
+	}
+
+	return nil
+}
+
+// TestAdminRequestMFA test that MFA can be provided for admin requests through request metadata,
+// both in the initial request with a call option or with a unary interceptor retry.
+func TestAdminRequestMFA(t *testing.T) {
+	t.Parallel()
+	ctx := context.Background()
+
+	srv := startMockServer(t, &mfaService{})
+
+	mfaTestResp := &proto.MFAAuthenticateResponse{
+		Response: &proto.MFAAuthenticateResponse_TOTP{
+			TOTP: &proto.TOTPResponse{
+				Code: otpTestCode,
+			},
+		},
+	}
+	t.Run("unary request", func(t *testing.T) {
+		t.Run("no MFA", func(t *testing.T) {
+			clt, err := New(ctx, srv.clientCfg())
+			require.NoError(t, err)
+			err = clt.CreateUser(ctx, &types.UserV2{})
+			require.Error(t, err)
```

Comment:
Is there something more specific we can assert about the error? Maybe:

```suggestion
			assert.True(t, trace.IsAccessDenied(err), "CreateUser returned err=%v (%T), want AccessDenied")
```

Same for others.